### PR TITLE
Updated loadHTML from jQuery to native Javascript

### PR DESCRIPTION
**Pull request by padmasreegade**

### CODE REVIEW

**Review by demiankatz — APPROVED** [review]
Thanks, @padmasreegade! This looks good to me and seems to be working as expected (I tested with both combined search and deferred recommendation modules, which seem to be the two main things that utilize this code). However, I'll leave it open until @crhallberg and/or @EreMaijala have a moment to take a look, just in case they have any observations that I've overlooked.

**Comment by padmasreegade** [comment]
> Thanks, @padmasreegade! This looks good to me and seems to be working as expected (I tested with both combined search and deferred recommendation modules, which seem to be the two main things that utilize this code). However, I'll leave it open until @crhallberg and/or @EreMaijala have a moment to take a look, just in case they have any observations that I've overlooked.

Thank you @demiankatz

File: themes/bootstrap3/js/common.js
Comment (demiankatz) [inline themes/bootstrap3/js/common.js]: One question: I notice that _element is likely to be a jQuery object in some situations. Does passing a jQuery object to querySelector work reliably, or should we perhaps either change the calling code or de-jQuery the object here?

In any case, I expect that #3283 is going to be merged very soon, which will greatly reduce the number of calls to the function by eliminating a large number of redundant template files. If it's worth changing the calls as well as the function, that should be much easier very soon...

Comment (demiankatz) [reply themes/bootstrap3/js/common.js]: As noted earlier, #3283 has been merged, so it may be worth merging the latest dev branch into this, especially if it's necessary to adjust the calls to the modified function.

Comment (padmasreegade) [reply themes/bootstrap3/js/common.js]: Will do it today @demiankatz

Comment (demiankatz) [reply themes/bootstrap3/js/common.js]: Thanks!

File: themes/bootstrap3/js/common.js
Comment (EreMaijala) [inline themes/bootstrap3/js/common.js]: The `onload` handler should be defined before `open()` is called. Otherwise there's at least in theory a possibility that the request completed before the onload handler is set up.

File: themes/bootstrap3/js/common.js
Comment (EreMaijala) [inline themes/bootstrap3/js/common.js]: I'm not sure this adds any useful information. But maybe we should update the target element's content to `VuFind.translate('error_occurred')` to indicate to the user that something went wront.

**Review by EreMaijala — CHANGES_REQUESTED** [review]
I'd suggest we standardize on the [Fetch API](https://developer.mozilla.org/en-US/docs/Web/API/Fetch_API). It uses promises, which avoids pitfalls with events. And we're already using it in record.js and search.js.

File: themes/bootstrap3/js/common.js
Comment (demiankatz) [inline themes/bootstrap3/js/common.js]: The previous version of the function had a success callback function as an argument. This feature is used by the loadHtml() call in combined-search.js, so without it, things are breaking.

In particular, it is causing this test failure:

```
1) VuFindTest\Mink\CombinedSearchTest::testCombinedSearchResultsAllAjax
Element not found: #combined_Solr____one a.title index 0
Failed asserting that null is of type "object".

/usr/local/vufind/module/VuFind/src/VuFindTest/Integration/MinkTestCase.php:408
/usr/local/vufind/module/VuFind/tests/integration-tests/src/VuFindTest/Mink/CombinedSearchTest.php:84
/usr/local/vufind/module/VuFind/tests/integration-tests/src/VuFindTest/Mink/CombinedSearchTest.php:144
/usr/local/vufind/module/VuFind/src/VuFindTest/Feature/AutoRetryTrait.php:104
```

When I test in my browser, I'm also seeing some other errors:

```
Request failed: TypeError: Window.fetch: HEAD or GET Request cannot have a body.
    loadHtml http://localhost/vufind_test/themes/bootstrap3/js/common.js?_=1705433595:249
    init http://localhost/vufind_test/themes/bootstrap3/js/combined-search.js?_=1705434206:5
    <anonymous> http://localhost/vufind_test/Combined/Results?lookfor=:256
    jQuery 2
common.js:263:17

Uncaught (in promise) TypeError: VuFind.translate is not a function
    loadHtml http://localhost/vufind_test/themes/bootstrap3/js/common.js?_=1705433595:264
2 common.js:264:36
```

I also pushed up 1bf1b357be3ac24b6bcf563a8d457f6edf835a53 to address my concerns about jQuery objects being passed to querySelector. However, because of the other issues noted above, I haven't been able to test whether these changes actually work correctly -- I just pushed them up in case they are helpful, but please feel free to revert my commit if they are not.

**Review by demiankatz — CHANGES_REQUESTED** [review]
Thanks for the progress here, @padmasreegade. Unfortunately, I seem to be running into some problems (possibly caused by code changes introduced after your initial refactoring). See below for details, and please let me know if I can do more to help!

File: themes/bootstrap3/js/common.js
Comment (padmasreegade) [reply themes/bootstrap3/js/common.js]: Working on it @demiankatz, will update code to use success callback and test.

**Review by demiankatz — CHANGES_REQUESTED** [review]
Thanks for the progress, @padmasreegade. Tests are passing again. Unfortunately, though, not all of the functionality is working right quite yet. The `include_recommendations_side` feature of combined.ini is no longer working on this branch, while it is working correctly in the dev branch. I am not entirely sure what has gone wrong here, so I am asking @maccabeelevine for a review in case he has a quick answer. If he is not sure, I can share details on how to set up a test scenario so we can do deeper debugging.

File: themes/bootstrap3/js/combined-search.js
Comment (maccabeelevine) [inline themes/bootstrap3/js/combined-search.js]: I used jQuery here originally only because it was [already being passed](https://github.com/vufind-org/vufind/pull/3233#discussion_r1402068771) in as a jQuery element, which you've fixed in this PR.  So you could replace this jQuery as well.  [This site](https://youmightnotneedjquery.com/) has one-line HTML replacements for hide(), parent(), and hasClass().

**Review by maccabeelevine — COMMENTED** [review]
Great to get rid of jQuery here!  I am not able to reproduce the problems with include_recommendations_side, but Demian may have another test case I'm missing.  I did include a comment below on improving it further.

**Comment by demiankatz** [comment]
@maccabeelevine, I'm still seeing the same problem as before. Here are my steps to reproduce:

1.) Spin up a standard test instance with phing startup
2.) Make these changes to combined.ini (I just used Search2 as a fake replacement for Summon/EDS to eliminate external dependencies, and added an instance of CatalogResults side recommendations):

```
diff --git a/config/vufind/combined.ini b/config/vufind/combined.ini
index b074ebb7ca..3bee050aeb 100644
--- a/config/vufind/combined.ini
+++ b/config/vufind/combined.ini
@@ -107,19 +107,20 @@ more_link = "More catalog results"
 ;advanced_link_bottom = "Advanced Search"
 ;limit = 10
 
-[Summon]
+[Search2:Summon]
 label = Summon
 ;sublabel = "full-text articles, e-books, electronic subscription resources"
 more_link = "More Summon results"
 ajax = true
 ;limit = 10
 
-[EDS]
+[Search2:EDS]
 label = EDS
 sublabel = "full-text articles, e-books, electronic subscription resources"
 more_link = "More EDS results"
 ajax = true
 ; limit = 10
+include_recommendations_side[] = CatalogResults

 ;[EIT]
 ;label = EIT
```

Now perform a combined search for "test". I get this in dev:

image

And I get this on the current branch:

image

File: themes/bootstrap3/js/common.js
Comment (maccabeelevine) [inline themes/bootstrap3/js/common.js]: Thanks @demiankatz for the steps to reproduce the include_recommendations_side problem, not sure why I could not do so before.  The problem is this line using `element.innerHTML` to set the incoming HTML content.  That HTML includes its own script to move itself to the sidebar, but setting innerHTML explicitly [does not run included scripts](https://www.w3.org/TR/2008/WD-html5-20080610/dom.html#innerhtml0).  I am honestly not sure why the jQuery html() function works since jQuery's docs say it uses .innerHTML, but it must be doing something else as well.  I did find a few potential solutions like [this ](https://stackoverflow.com/posts/47614491/revisions)but alternately it might be better to move that script out of the included HTML in the first place, which I can look into if needed.

**Review by demiankatz — CHANGES_REQUESTED** [review]
Thanks, @padmasreegade, your simplification of the combined-search.js changes looks good (though if you haven't already, it might be worth reviewing/asking @crhallberg whether setting display to none is our conventional way of hiding things, or if we should instead be setting a hidden class instead).

The problem with scripts not executing in loadHTML is still a problem, though. Please see my suggested changes to combined.ini for a way to demonstrate the problem, and @maccabeelevine's comments about a possible solution for more detail -- and let me know if you need additional clarification!

**Comment by crhallberg** [comment]
> display: none vs hidden class

I don't believe we have a set way of doing it. Setting display to none is direct, performant, and the non-jQuery equivalent of .hide() so no concerns here!

**Comment by padmasreegade** [comment]
Made changes to use DOM parsing technique and catalog results can now be seen. 
image

File: themes/bootstrap3/js/common.js
Comment (demiankatz) [inline themes/bootstrap3/js/common.js]: It would probably be a good idea to add a comment here to explain the purpose of this code. For example, something like:

```
// If the loaded HTML includes script tags, extra work is needed to ensure that they execute:
```

**Review by demiankatz — CHANGES_REQUESTED** [review]
Thanks, @padmasreegade, this is working for me and looks good. Just one small suggestion below.

I'm also interested in whether @crhallberg and @EreMaijala have any remaining concerns before this is merged.

File: themes/bootstrap3/js/combined-search.js
Comment (EreMaijala) [inline themes/bootstrap3/js/combined-search.js]: I think this should be `        while (parent && parent.classList.contains('js-hide-if-empty')) {`. We may not currently have a case where not having the check would be a problem, but I'd rather account for the fact that parentNode may return null.

File: themes/bootstrap3/js/combined-search.js
Comment (EreMaijala) [inline themes/bootstrap3/js/combined-search.js]: These is still jQuery and need to be updated.

File: themes/bootstrap3/js/common.js
Comment (EreMaijala) [inline themes/bootstrap3/js/common.js]: No need for `Array.from()` here. querySelectorAll returns NodeList that also has the forEach method.

File: themes/bootstrap3/js/common.js
Comment (EreMaijala) [inline themes/bootstrap3/js/common.js]: I don't think it's necessary to copy the attributes one by one. Actually, you could just copy the whole script as is and just replace the nonce attribute before replaceChild below. Also, the code needs to check script type before modifying it. We have some that are not JavaScript. See https://github.com/vufind-org/vufind/blob/cbd39fcd384bec72d9a798cb97a35567b2880438/themes/bootstrap3/js/search.js#L335 and https://github.com/vufind-org/vufind/blob/cbd39fcd384bec72d9a798cb97a35567b2880438/themes/bootstrap3/js/search.js#L351 for example. Perhaps we could make it into a shared helper method?

File: themes/bootstrap3/js/common.js
Comment (EreMaijala) [inline themes/bootstrap3/js/common.js]: If we handle nonce in setInnerHtml, updateCspNonce should be removed here.

**Review by EreMaijala — CHANGES_REQUESTED** [review]

File: themes/bootstrap3/js/combined-search.js
Comment (demiankatz) [inline themes/bootstrap3/js/combined-search.js]: @padmasreegade, this seems to have reverted to using jQuery. Was that intentional? (In any case, I assume work is currently in progress, so I won't look again until you let me know you're ready for another review. :-) )

**Comment by demiankatz** [comment]
Thanks, @padmasreegade! Looks to me like all of @EreMaijala's feedback has been addressed except for the refactoring suggestion (which would probably be better addressed in a separate PR anyway). The test suite is still passing for me.

And regarding the refactoring, what exactly did you have in mind, @EreMaijala? Something like this?

```
function forEachJavascript(element, callback) {
  element.querySelectorAll('script').forEach((el) => {
  const type = el.getAttribute('type');
  if (!type || 'text/javascript' === type) {
    callback(el);
  }
});
```

**Review by demiankatz — APPROVED** [review]
I'm approving, but will wait for @EreMaijala's approval before merging.

**Comment by EreMaijala** [comment]
@demiankatz Actually, `setInnerHtml` is now pretty good. We just need `setOuterHtml` to accompany it, and search.js could use them, I think. I can create a follow-up PR.

**Review by EreMaijala — APPROVED** [review]

**Comment by demiankatz** [comment]
Excellent! Thanks for the persistence on this surprisingly challenging task, @padmasreegade, and for the very thorough reviewing, @EreMaijala. I'll merge this now and look forward to @EreMaijala's follow-up when time permits.

**Pull request merged**
